security-csrf/src/main/java/io/micronaut/security/csrf/filter/CsrfFilter.java

```diff
@@ -0,0 +1,224 @@
+    private static final Logger LOG = LoggerFactory.getLogger(CsrfFilter.class);
+    private static final Mono<Optional<HttpResponse<?>>> PROCEED = Mono.just(Optional.empty());
+    private final List<FutureCsrfTokenResolver<HttpRequest<?>>> futureCsrfTokenResolvers;
+    private final List<CsrfTokenResolver<HttpRequest<?>>> csrfTokenResolvers;
+    private final CsrfTokenValidator<HttpRequest<?>> csrfTokenValidator;
+    private final ExceptionHandler<AuthorizationException, MutableHttpResponse<?>> exceptionHandler;
+    private final CsrfFilterConfiguration csrfFilterConfiguration;
+
+    CsrfFilter(CsrfFilterConfiguration csrfFilterConfiguration,
+               List<FutureCsrfTokenResolver<HttpRequest<?>>> futureCsrfTokenResolvers,
+               List<CsrfTokenResolver<HttpRequest<?>>> csrfTokenResolvers,
+               CsrfTokenValidator<HttpRequest<?>> csrfTokenValidator,
+               ExceptionHandler<AuthorizationException, MutableHttpResponse<?>> exceptionHandler) {
+        this.csrfTokenResolvers = csrfTokenResolvers;
+        this.futureCsrfTokenResolvers = futureCsrfTokenResolvers.isEmpty()
+                ? futureCsrfTokenResolvers
+                : FutureCsrfTokenResolver.of(csrfTokenResolvers, futureCsrfTokenResolvers);
+        this.csrfTokenValidator = csrfTokenValidator;
+        this.exceptionHandler = exceptionHandler;
+        this.csrfFilterConfiguration = csrfFilterConfiguration;
+    }
+
+    @RequestFilter
+    @Nullable
+    public Publisher<Optional<HttpResponse<?>>> csrfFilter(@NonNull HttpRequest<?> request) {
+        if (!shouldTheFilterProcessTheRequestAccordingToTheUriMatch(request)) {
+            return PROCEED;
+        }
+        if (!shouldTheFilterProcessTheRequestAccordingToTheHttpMethod(request)) {
+            return PROCEED;
+        }
+        if (!shouldTheFilterProcessTheRequestAccordingToTheContentType(request)) {
+            return PROCEED;
+        }
+        return futureCsrfTokenResolvers.isEmpty()
+                ? imperativeFilter(request)
+                : reactiveFilter(request);
+    }
+
+    boolean shouldTheFilterProcessTheRequestAccordingToTheUriMatch(HttpRequest<?> request) {
+        RouteMatch<?> routeMatch = request.getAttribute(HttpAttributes.ROUTE_MATCH, RouteMatch.class).orElse(null);
+        if (routeMatch instanceof UriRouteMatch<?, ?> uriRouteMatch) {
+            return shouldTheFilterProcessTheRequestAccordingToTheUriMatch(uriRouteMatch);
+        }
+        return true;
+    }
+
+    boolean shouldTheFilterProcessTheRequestAccordingToTheUriMatch(UriRouteMatch<?, ?> uriRouteMatch) {
+        return shouldTheFilterProcessTheRequestAccordingToTheUriMatch(uriRouteMatch.getUri());
+    }
+
+    boolean shouldTheFilterProcessTheRequestAccordingToTheUriMatch(String uri) {
+        boolean matches = PathMatcher.REGEX.matches(csrfFilterConfiguration.getRegexPattern(), uri);
+        if (!matches) {
+            if (LOG.isTraceEnabled()) {
```

Comment:
Looks like you use a lot of this check maybe create a static field

Reply:
The log levels be changed dynamically via the logger endpoint for example. If I move the logic to a static field e.g.

```
private final static boolean isTraceEnabled = LOG.isTracedEnabled()
```

That would be set in stone and changing the log levels will not work.